File: python/production/utils/location.py
from production.utils.face_handler import Face_handler

class Location:

    def containsFace(self,p_face):
        if p_face == self.c_face0 or p_face == self.c_face1 or ( (not self.c_isEdge) and (p_face == self.c_face2)):
            return True
        else:
            return False
    

    def get_copy(self):
        if self.isEdge():
            return Location(self.c_face0, self.c_face1)
        else:
            return Location(self.c_face0, self.c_face1, self.c_face2)
    

    def __init__(self,p_face0,p_face1,p_face2=None):
        if p_face2 == None:
            self.location2(p_face0,p_face1)
        else:
            self.c_isEdge = False
            self.c_face0 = p_face0
            self.c_face1 = p_face1
            self.c_face2 = p_face2
            # bubble sort to keep order within faces		
            if self.c_face0 > self.c_face1:
            
                l_tmp = self.c_face1
                self.c_face1 = self.c_face0
                self.c_face0 = l_tmp
            
            if self.c_face1 > self.c_face2:
            
                l_tmp = self.c_face2
                self.c_face2 = self.c_face1
                self.c_face1 = l_tmp
            
            if self.c_face0 > self.c_face1:
            
                l_tmp = self.c_face1
                self.c_face1 = self.c_face0
                self.c_face0 = l_tmp
        

    

    def location2(self,p_face0,p_face1):
        self.c_isEdge = True
        self.c_face2 = None
        self.c_face0 = p_face0
        self.c_face1 = p_face1
        if p_face0 > p_face1:
        
            self.c_face1 = p_face0
            self.c_face0 = p_face1
        
        else:
        
            self.c_face0 = p_face0
            self.c_face1 = p_face1
        
    

    def isEdge(self):
        return self.c_isEdge
    

    def getFace0(self):
        return self.c_face0
    

    def getFace1(self):
        return self.c_face1
    

    def getFloor(self):
        if self.equals(Location("TOP", "LEFT", "FRONT")):
            return 3
        if self.equals(Location("TOP", "LEFT", "BACK")):
            return 3
        if self.equals(Location("TOP", "RIGHT", "FRONT")):
            return 3
        if self.equals(Location("TOP", "RIGHT", "BACK")):
            return 3
        if self.equals(Location("BOTTOM", "LEFT", "FRONT")):
            return 1
        if self.equals(Location("BOTTOM", "LEFT", "BACK")):
            return 1
        if self.equals(Location("BOTTOM", "RIGHT", "FRONT")):
            return 1
        if self.equals(Location("BOTTOM", "RIGHT", "BACK")):
            return 1

        if self.equals(Location("TOP", "FRONT")):
            return 3
        if self.equals(Location("TOP", "BACK")):
            return 3
        if self.equals(Location("TOP", "LEFT")):
            return 3
        if self.equals(Location("TOP", "RIGHT")):
            return 3

        if self.equals(Location("FRONT", "LEFT")):
            return 2
        if self.equals(Location("FRONT", "RIGHT")):
            return 2
        if self.equals(Location("BACK", "LEFT")):
            return 2
        if self.equals(Location("BACK", "RIGHT")):
            return 2

        if self.equals(Location("BOTTOM", "LEFT")):
            return 1
        if self.equals(Location("BOTTOM", "RIGHT")):
            return 1
        if self.equals(Location("FRONT", "BOTTOM")):
            return 1
        if self.equals(Location("BACK", "BOTTOM")):
            return 1
        return 0
    

    def getFace2(self):
        if self.isEdge():
            return None
        else:
            return self.c_face2
    

    #	getValue(self):
    #		return (getFace0() * 1 + getFace1() * 6 + getFace2() * 36 + isEdge() * 216)
    #	


    
    def equals(self, p_location):
        return ((self.c_face0 == p_location.getFace0()) and
                (self.c_face1 == p_location.getFace1()) and
                (self.c_face2 == p_location.getFace2()) and
                (self.c_isEdge == p_location.isEdge()))
    


    def getString(self):
        if self.c_isEdge:
            return Face_handler.getCharValue(self.c_face0) + ", " + Face_handler.getCharValue(self.c_face1)
        else:
        
            return Face_handler.getCharValue(self.c_face0) + ", " + Face_handler.getCharValue(self.c_face1) + ", " + Face_handler.getCharValue(self.c_face2)
        
    

File: csharp/Production/utils/Location.cs
using System;


namespace utils
{

    public class Location
    {
        Boolean c_isEdge;
        private Face c_face0;
        private Face c_face1;
        private Face c_face2;

        Location()
        {
        }

        public Boolean containsFace(Face p_face)
        {
            if (p_face == c_face0 || p_face == c_face1 || (!c_isEdge && (p_face == c_face2)))
                return true;
            else
                return false;
        }

        public Location getCopy()
        {
            if (isEdge())
                return new Location(c_face0, c_face1);
            else
                return new Location(c_face0, c_face1, c_face2);
        }

        public Location(Face p_face0, Face p_face1, Face p_face2)
        {
            Face l_tmp;
            c_isEdge = false;
            c_face0 = p_face0;
            c_face1 = p_face1;
            c_face2 = p_face2;
            // bubble sort to keep order within faces		
            if ((int)c_face0 > (int)c_face1)
            {
                l_tmp = c_face1;
                c_face1 = c_face0;
                c_face0 = l_tmp;
            }
            if ((int)c_face1 > (int)c_face2)
            {
                l_tmp = c_face2;
                c_face2 = c_face1;
                c_face1 = l_tmp;
            }
            if ((int)c_face0 > (int)c_face1)
            {
                l_tmp = c_face1;
                c_face1 = c_face0;
                c_face0 = l_tmp;
            }

        }

        public Location(Face p_face0, Face p_face1)
        {
            c_isEdge = true;
            //       c_face2 = Face.NOTDEFINED;
            c_face0 = p_face0;
            c_face1 = p_face1;
            if ((int)p_face0 > (int)p_face1)
            {
                c_face1 = p_face0;
                c_face0 = p_face1;
            }
            else
            {
                c_face0 = p_face0;
                c_face1 = p_face1;
            }
        }

        public Boolean isEdge()
        {
            return c_isEdge;
        }

        public Face getFace0()
        {
            return c_face0;
        }

        public Face getFace1()
        {
            return c_face1;
        }

        public int getFloor()
        {
            if (this.equals(new Location(Face.TOP, Face.LEFT, Face.FRONT))) return 3;
            if (this.equals(new Location(Face.TOP, Face.LEFT, Face.BACK))) return 3;
            if (this.equals(new Location(Face.TOP, Face.RIGHT, Face.FRONT))) return 3;
            if (this.equals(new Location(Face.TOP, Face.RIGHT, Face.BACK))) return 3;
            if (this.equals(new Location(Face.BOTTOM, Face.LEFT, Face.FRONT))) return 1;
            if (this.equals(new Location(Face.BOTTOM, Face.LEFT, Face.BACK))) return 1;
            if (this.equals(new Location(Face.BOTTOM, Face.RIGHT, Face.FRONT))) return 1;
            if (this.equals(new Location(Face.BOTTOM, Face.RIGHT, Face.BACK))) return 1;

            if (this.equals(new Location(Face.TOP, Face.FRONT))) return 3;
            if (this.equals(new Location(Face.TOP, Face.BACK))) return 3;
            if (this.equals(new Location(Face.TOP, Face.LEFT))) return 3;
            if (this.equals(new Location(Face.TOP, Face.RIGHT))) return 3;

            if (this.equals(new Location(Face.FRONT, Face.LEFT))) return 2;
            if (this.equals(new Location(Face.FRONT, Face.RIGHT))) return 2;
            if (this.equals(new Location(Face.BACK, Face.LEFT))) return 2;
            if (this.equals(new Location(Face.BACK, Face.RIGHT))) return 2;

            if (this.equals(new Location(Face.BOTTOM, Face.LEFT))) return 1;
            if (this.equals(new Location(Face.BOTTOM, Face.RIGHT))) return 1;
            if (this.equals(new Location(Face.FRONT, Face.BOTTOM))) return 1;
            if (this.equals(new Location(Face.BACK, Face.BOTTOM))) return 1;
            return 0;
        }

        public Face getFace2()
        {
            //        if (isEdge())
            //            return Face.NOTDEFINED;
            //        else
            return c_face2;
        }

        //	int getValue() {
        //		return (getFace0() * 1 + getFace1() * 6 + getFace2() * 36 + isEdge() * 216);
        //	}

        public Boolean equals(Location p_location)
        {
            return ((c_face0 == p_location.c_face0) &&
                    (c_face1 == p_location.c_face1) &&
                    (c_face2 == p_location.c_face2) &&
                    (c_isEdge == p_location.c_isEdge));
        }


        public String getString()
        {
            if (c_isEdge)
                return String.Format("{0}, {1}", FaceHandler.getCharValue(c_face0), FaceHandler.getCharValue(c_face1));
            else
            {
                return String.Format("{0}, {1}, {2}", FaceHandler.getCharValue(c_face0), FaceHandler.getCharValue(c_face1), FaceHandler.getCharValue(c_face2));
            }
        }

    }
}